casper/src/main/scala/coop/rchain/casper/TuranOracle.scala

```diff
@@ -0,0 +1,103 @@
+package coop.rchain.casper
+
+import com.google.protobuf.ByteString
+import coop.rchain.casper.protocol.{BlockMessage, Bond, Justification}
+import util._
+
+import scala.collection
+
+/*
+ * Implementation inspired by Ethereum's CBC casper simulator's Turan oracle implementation.
+ *
+ * https://github.com/ethereum/cbc-casper/blob/0.2.0/casper/safety_oracles/turan_oracle.py
+ *
+ * "If nodes in an e-clique see each other agreeing on e and can't see each other disagreeing on e,
+ * then there does not exist any new message from inside the clique that will cause them to assign
+ * lower scores to e. Further, if the clique has more than half of the validators by weight,
+ * then no messages external to the clique can raise the scores these validators assign to
+ * a competing estimate to be higher than the score they assign to e."
+ *
+ * - From https://github.com/ethereum/research/blob/master/papers/CasperTFG/CasperTFG.pdf
+ *
+ * That is unless there are equivocations.
+ * The fault tolerance threshold is a subjective value that the user sets to "secretly" state that they
+ * tolerate up to fault_tolerance_threshold fraction of the total weight to equivocate.
+ *
+ * In the extreme case when your fault tolerance threshold is 1,
+ * all validators must be part of the clique that supports the estimate in order to state that it is finalized.
+ * If all validators are indeed part of the clique, min_max_clique_weight will hopefully be equal to total_weight
+ * and is_safe will reduce to total_weight >= total_weight and evaluate to true.
+ */
+class TuranOracle(blocks: collection.Map[ByteString, BlockMessage],
+                  estimate: BlockMessage,
```

Comment:
Does it make sense for `estimate` to be a parameter in the class constructor as opposed to an argument of the `isSafe` method? Just thinking it might be more efficient to be able to use the same class instance to test multiple blocks.